device/core/test/_internal_client_test.js

```diff
@@ -278,15 +278,6 @@ var ModuleClient = require('../lib/module_client').ModuleClient;
         });
       });
 
-      it('calls the callback immediately if the client is already disconnected', function (testCallback) {
```

Comment:
This unit test looks incorrect in the first place. The close method looks like this:
`  close(closeCallback?: Callback<results.Disconnected>): Promise<results.Disconnected> | void {
    return callbackToPromise((_callback) => {
      this._closeTransport((err, result) => {
        safeCallback(_callback, err, result);
      });
    }, closeCallback);
  }`

which calls _closeTransport that looks like this:

`  private _closeTransport(closeCallback: (err?: Error, result?: any) => void): void {
    const onDisconnected = (err?: Error, result?: any): void => {
      /*Codes_SRS_NODE_INTERNAL_CLIENT_16_056: [The `close` method shall not throw if the `closeCallback` is not passed.]*/
      /*Codes_SRS_NODE_INTERNAL_CLIENT_16_055: [The `close` method shall call the `closeCallback` function when done with either a single Error object if it failed or null and a results.Disconnected object if successful.]*/
      safeCallback(closeCallback, err, result);
    };

    /*Codes_SRS_NODE_INTERNAL_CLIENT_16_046: [The `close` method shall remove the listener that has been attached to the transport `disconnect` event.]*/
    this._transport.removeListener('disconnect', this._disconnectHandler);
    /*Codes_SRS_NODE_INTERNAL_CLIENT_16_001: [The `close` function shall call the transport's `disconnect` function if it exists.]*/
    this._transport.disconnect((disconnectError, disconnectResult) => {
      onDisconnected(disconnectError, disconnectResult);
    });
  }
}`

The code doesn't check if the client is already disconnected, it calls `this._transport.disconnect` in all paths. The reason why it worked was because `testCallback()` was passed instead of `testCallback` which caused unit test to exit immediately and not pass any callback to the method. Please correct me if I'm wrong.

Reply:
Done.

---
In reply to: [218877724](https://github.com/Azure/azure-iot-sdk-node/pull/372#discussion_r218877724) [](ancestors = 218877724)

Reply:
Won't fix.

---
In reply to: [220019190](https://github.com/Azure/azure-iot-sdk-node/pull/372#discussion_r220019190) [](ancestors = 220019190)